sips/sip-x/sip-x-authentication-protocol.md

```diff
@@ -0,0 +1,397 @@
+| redirect_uri           | string | The url that should receive the authentication response                                                                                   |
+| do_not_include_profile | bool   |                                                                                                                                           |
+| supports_hub_url       | bool   |                                                                                                                                           |
+| scopes                 | array  | list of strings specifying the requested access to the user's account. See Appendix A for full list of scopes                             |
+| state                  | any    | value to be echoed in the auth response                                                                                                   |
+| version                | string | must be "2.0.0"                                                                                                                           |
+
+### Verification
+
+Authenticators should verify that the request has the following properties:
+
+- expiration date (`exp`) is not in the past
+- issuance date (`iat`) is in the past
+- public keys length is 1
+- public key (`public_keys[0]`) is same as the signer's key
+- public key's stacks address is the same as the issuer
+- manifest url is same origin as the app domain
+- redirect url is same origin as the app domain
+
+## User authorization
+
+The authenticator manages the user' private keys. The protocol requires that keys are created from a deterministic wallet using [BIP-39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki).
+
+The authenticator can offer the user a list of accounts. Each account corresponds to a change of the key derivation path by 1. This SIP does not specify how the wallet determines which accounts should be presented. Only requirement is that one account is selected. The selected index shall be called account index `n`.
+
+Once the account index has been fixed, authenticator has to do the following operations:
+
+- Update public profile (depending on requested scopes)
+- Create authentication response
+
+## Application Private Key
+
+If the authentication request contains the scope `storage_write`, then the authenticator must create a private key that is specific for the requesting application. It must use the following algorithm:
+
+1. Wallet salt: create the sha256 hash of the hex representation of the public key of derivation path (`m/888'/0'`)
+1. Create sha256 hash of concatinated string of the `domain_name` from the request and the hex representation of the wallet salt.
+1. Hash code: Create a hash code (as defined in Java and Javascript) from the hex representation of the hash, then apply bitwise `and` to the hash code and 0x7fffffff.
```

Comment:
This is highly prone to collisions, and generally a bad implementation. Unfortunately, it has been used for years. I have created what I think is a better alternative [here in micro-stacks](https://github.com/fungible-systems/micro-stacks/blob/8d9e2493e0ed751bb64a7a7cd574b692c068f8d7/src/wallet-sdk/account/get-app-private-key.ts#L22-L37).

```ts
  // Rather than using the `hashCode` function to derive an index for the child
  // which makes it so there's a unreasonable high probability that someone can
  // log into a new app (website) and it accidentally shares data with another one
  //
  // we want to use the `appDomain` + `salt` to generate a sha256 hash that is
  // then used as the chaincode for the new app keychain
  const chainCode = hashSha256(utf8ToBytes(`${appDomain}${account.salt}`));
  const rootNode = HDKeychain.fromBase58(account.appsKey);
  const appKeychain = await HDKeychain.fromPrivateKey(rootNode.privateKey, chainCode).deriveChild(
    0,
    true
  );
  ```

With the formalization of this standard in a SIP, would it make sense to explore this (or another) alternative?

Reply:
cc @zone117x @kantai for any thoughts you might have on this

Reply:
Interesting. We should fix that!